// @stryker-mutator/tap-runner@9.3.0/src-generated/tap-runner-options.ts
/**
 * This file was automatically generated by generate-json-schema-to-ts.js.
 * DO NOT MODIFY IT BY HAND. Instead, modify the source file JSON file: tap-runner-options.json,
 * and run 'npm run generate' from monorepo base directory.
 */

export interface StrykerTapRunnerOptions {
  tap: TapRunnerOptions;
}
/**
 * Configuration for @stryker-mutator/tap-runner
 */
export interface TapRunnerOptions {
  /**
   * Specify glob expressions to test files to run. Defaults to ["{** /@(test|tests|__test__|__tests__)/**,** /*.@(test|tests|spec)}.@(cjs|mjs|js|jsx|ts|tsx)"]
   */
  testFiles: string[];
  /**
   * Specify additional node arguments to be passed when running your test files. The following template variables could be used: `{{hookFile}}` and `{{testFile}}`. Note that stryker will always pass an additional `--require` flag to enable coverage analysis. In addition stryker will also prepend `-r {{hookFile}}` and prepend `{{testFile}}` if not found the template variable is not found.
   */
  nodeArgs: string[];
  /**
   * Force a running test process to hard-exit after first bail test failure when disableBail is false.
   */
  forceBail: boolean;
}
